src/foam/nanos/auth/GiveUpdatePermissionDAO.java

```diff
@@ -0,0 +1,51 @@
+/**
+ * @license
+ * Copyright 2018 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package foam.nanos.auth;
+
+import foam.core.FObject;
+import foam.core.X;
+import foam.dao.DAO;
+import foam.dao.ProxyDAO;
+
+/**
+ * When a user creates a new group, give the group they're in permission to
+ * update the new group, which will also allow them to put other users in this
+ * group.
+ */
+public class GiveUpdatePermissionDAO extends ProxyDAO {
+  public GiveUpdatePermissionDAO(X x, DAO delegate) {
+    super(x, delegate);
+  }
+
+  @Override
+  public FObject put_(X x, FObject obj) {
+    Group toPut = (Group) obj;
+    Group existing = (Group) getDelegate().find_(x, toPut.getId());
+    AuthService auth = (AuthService) x.get("auth");
+
+    if ( existing == null && ! auth.check(x, "group.update." + toPut.getId()) ) {
```

Comment:
The way we addressed this in the past was by using wildcards for groups. For example, if we wanted some company admin to have the ability to create new groups we might give them the group.update.acme.* permission. Then they would create all of their groups as acme.operator, acme.admin, acme.accountant.level1, etc. This helps keep the group namespace orderly when being used by multiple companies. Also, the problem with the above approach is if I had an Acme.admin group who creates an acme.junioradmin group who creates acme.something else, then the original Acme.admin wouldn't have access (directly) to the new acme.something group, but by using hierarchical groups they would.

Reply:
Say we give some admin group for Acme the `"group.update.acme.*"` permission, but they create a group just called `"employee"` instead of `"acme.employee"` like we wanted them to. They still don't have permission to that group because they only have `"group.update.acme.*"`.

I guess we could solve that by making the `group.create` permission more fine-grained as well, something like `group.create.acme`. Then they would only be able to create groups prefixed with `"acme."`. Does that work with what you had in mind?

Reply:
Yes, the system just wouldn't let them create groups with just the 'employee' id. The Auth decorator could even just imply the update permission if you had the create permission.